### ISSUE TITLE
Module Development Task:  Automated Blacklisting

### ISSUE THREAD

**Issue by CPSTL** [PolymathNetwork/polymath-core#272]
# Module Development Task
 # Automated Blacklisting

Module Type | Bounty | Completion deadline
-- | -- | --
Transfer Manager | Approx. 5, 000 POLY | 21 days to complete once task is confirmed by external developer(s).

# Short Summary
This module will allow approved staff to create and establish an automated blacklist to prevent Insiders from selling thus de-risking price swings during time sensitive cases like earning calls. 

# Bounty Requirements
1. Module specs are fulfilled
2. Module is implemented using appropriate module interface
3. Module is tested with >95% branch coverage
4. Module is delivered before deadline
5. Module does not present any bugs.
6. 21 days to complete once task work has begun.

**IMPORTANT:** Please work from the development-1.5.0 branch instead of master.

# Suggested Implementation

```
Struct BlacklistDetails {
	uint256 startDate;
    	uint256 endDate;
	uint256 repeatPeriodInDays;
}
/// use to hold the different blacklist corresponds to name of blacklist type 
     mapping(bytes32 => BlacklistDetails) blacklists;
/// use to hold the blackListed address corresponds to the blacklist type
     mapping(address => bytes32) inverstorToBlacklist;
/// use to get list of the addresses for a particular blacklist
     mapping(bytes32 => address[]) blacklistToAddress;


// If Address is blackListed and its blacklist details lie within the current time then the transfer doesn’t happen should return INVALID
function verifyTransfer(address _from, address _to, uint256 /* _amount*/, bool /* _isTransfer */) public returns(Result) {
      if (investorToBlackList[_from] != bytes32(0)) {
	/// Add the logic for repeat periods
          if (blacklists[investorToBlackList[_from]].endDate >= now && blacklists[investorToBlackList[_from]].startDate <= now) {
return Result.INVALID;
}
       } 
       return Result.NA;
}

function addBlackListType(uint256 _startDate, uint256 _endDate, bytes32 _name, uint256 _repeatPeriodInDays);
function addInvestorToBlackList(address _investor, bytes32 _blackListName);
```
# Module Details Spec (Wireframe) 

**Wireframe Link:** https://www.lucidchart.com/invitations/accept/c544a325-e74d-4e54-a360-71805a9f10ce

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 5000.0 POLY (690.0 USD @ $0.14/POLY)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $21,413.49 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 3 years, 4 months ago.
Please review their action plans below:


**1) [moodysalem](https://gitcoin.co/moodysalem) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/polymathnetwork/polymath-core/272/1246?mutate_worker_action=approve&worker=moodysalem) | [reject worker](https://gitcoin.co/issue/polymathnetwork/polymath-core/272/1246?mutate_worker_action=reject&worker=moodysalem))_.**

I wrote the bounty0x ICO contracts and graduated from the Consensys academy and built ethercast.io and implementing this interface and testing looks relatively simple. I can start on Monday.
**2) [vpanwar821](https://gitcoin.co/vpanwar821) has been approved to start work.**

This is my action plan:
1) Will go through the requirement.
2) Develop the code according to the requirement with the 100% test coverage.
3) Mythrill sanity check

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/polymathnetwork/polymath-core/272/1246).

**Comment by satyamakgec** [comment]
@kpulkit29 All the specs are listed [here](https://www.lucidchart.com/invitations/accept/c544a325-e74d-4e54-a360-71805a9f10ce) and you also have the suggested implementation written in the issue. If you still have some confusion then you can ask queries here (issue thread) or on our gitter [channnel](https://gitter.im/PolymathNetwork/Lobby)

**Comment by kpulkit29** [comment]
Does the automated blacklist entry need to be made in the contract??

**Comment by kpulkit29** [comment]
I mean the part where approved staff click  enables the blacklist.I guess
this is how it needs to be done store the blacklist details in the
contract.For the staff to see the details the entered blacklist details are
pulled out from the contract.Then eth address can be assigned through a
contract function to a particular blacklist.

**Comment by pabloruiz55** [comment]
@kpulkit29 yes, one way to do it would be to have a set of rules, for example:
Start date: 10000, Duration: 1000, interval: 1500. This would mean that for an address following this rule transactions out would fail from timestamp 10,000 to 11,000; 12,500 to 13,500; 15,000 to 16,000, and so on.

Then, addresses could be "added" so they follow one of these rule sets.

Both rule sets and addresses would be added either by the issuer (SecurityToken owner) or someone they delegate (check other modules to see how they work with the permissions manager).

**Comment by kpulkit29** [comment]
What does interval signify here??I would like to start work on this if
approved on gitcoin explorer.

**Comment by pabloruiz55** [comment]
It means how often the start/end repeats. As per my example, you would want to block transfers for a group during 15 days, every 4 months.

**Comment by kpulkit29** [comment]
Sir I would be able to start work if you can approve

**Comment by CPSTL** [comment]
@kpulkit29 Approved :)

**Comment by kpulkit29** [comment]
ParserError: Expected token Comma got 'Identifier'
        (bytes32 price, bool valid) = Medianizer(makerDAO).peek();

**Comment by kpulkit29** [comment]
Getting this error on truffle compile not able to get the repo running

**Comment by pabloruiz55** [comment]
@kpulkit29 which branch are you woking on? Truffle / solc versions?

**Comment by kpulkit29** [comment]
truffle 4.1.18
solc, Version: 0.4.24+commit.e67f0147.Darwin.appleclang

Should I only try running at dev-1.5 branch

**Comment by satyamakgec** [comment]
@kpulkit29 I don't think so that compilation error exists in any of our branches (Please tell me which branch you take as the base branch so I can cross-check),  Please check your environment setup and follow these [environment requirements](https://github.com/PolymathNetwork/polymath-core#package-version-requirements-for-your-machine). And yes you only need to work corresponds to `development-1.5.0` branch

**Comment by kpulkit29** [comment]
pragma solidity ^0.4.24;
/**
 * @title AutomateBlacklist
 * @dev to automate blacklist process
 */
contract AutomateBlacklist {
    
    /**
     * @dev structure for the blacklist period
     */
     struct BlacklistDetails {
         uint256 startDate;
         uint256 endDate;
         uint256 repeats;
     }
     /**
      *@dev to check if the maaping of investor exists 
      */
      modifier isPresent(address _investor) {
          require(bytes(inverstorToBlacklist[_investor]).length!=0);
          _;
      } 
     
  /// use to hold the different blacklist corresponds to name of blacklist type 
     mapping(string => BlacklistDetails) blacklists;
/// use to hold the blackListed address corresponds to the blacklist type
     mapping(address => string) inverstorToBlacklist;
/// use to get list of the addresses for a particular blacklist
     mapping(string => address[]) blacklistToAddress;
    
    /**
     * @dev to check if transfer can be made
     */
     function canBeTransferred(address _investor) isPresent(_investor) public returns(bool) {
         string storage typeOfBlacklist=inverstorToBlacklist[_investor];
         if(now>blacklists[typeOfBlacklist].endDate){
             blacklists[typeOfBlacklist].startDate+=blacklists[typeOfBlacklist].repeats;
             blacklists[typeOfBlacklist].endDate+=blacklists[typeOfBlacklist].repeats;
             //checking the updated timeline
              if(blacklists[typeOfBlacklist].startDate>=now&&blacklists[typeOfBlacklist].endDate<=now){
            return false;
        }
         }
       
        return true;
     }
     /**
      * @dev to add the new blacklist entry
      */
      function addEntry(string _type,address _investor,uint256 _start,uint256 _end,uint256 _repeat) public {
          blacklists[_type]=BlacklistDetails(_start,_end,_repeat);
          inverstorToBlacklist[_investor]=_type;
          blacklistToAddress[_type].push(_investor);
      }
     /**
      * @dev to edit the existing blacklist entry
      */
      function editEntry(string _type,address _investor,uint256 _start,uint256 _end,uint256 _repeat) isPresent(_investor) public {
          blacklists[_type]=BlacklistDetails(_start,_end,_repeat);
          inverstorToBlacklist[_investor]=_type;
      }
     

}

**Comment by kpulkit29** [comment]
`pragma solidity ^0.4.24;
/**
 * @title AutomateBlacklist
 * @dev to automate blacklist process
 */
contract AutomateBlacklist {
    
    /**
     * @dev structure for the blacklist period
     */
     struct BlacklistDetails {
         uint256 startDate;
         uint256 endDate;
         uint256 repeats;
     }
     /**
      *@dev to check if the maaping of investor exists 
      */
      modifier isPresent(address _investor) {
          require(bytes(inverstorToBlacklist[_investor]).length!=0);
          _;
      } 
     
  /// use to hold the different blacklist corresponds to name of blacklist type 
     mapping(string => BlacklistDetails) blacklists;
/// use to hold the blackListed address corresponds to the blacklist type
     mapping(address => string) inverstorToBlacklist;
/// use to get list of the addresses for a particular blacklist
     mapping(string => address[]) blacklistToAddress;
    
    /**
     * @dev to check if transfer can be made
     */
     function canBeTransferred(address _investor) isPresent(_investor) public returns(bool) {
         string storage typeOfBlacklist=inverstorToBlacklist[_investor];
         if(now>blacklists[typeOfBlacklist].endDate){
             blacklists[typeOfBlacklist].startDate+=blacklists[typeOfBlacklist].repeats;
             blacklists[typeOfBlacklist].endDate+=blacklists[typeOfBlacklist].repeats;
             //checking the updated timeline
              if(blacklists[typeOfBlacklist].startDate>=now&&blacklists[typeOfBlacklist].endDate<=now){
            return false;
        }
         }
       
        return true;
     }
     /**
      * @dev to add the new blacklist entry
      */
      function addEntry(string _type,address _investor,uint256 _start,uint256 _end,uint256 _repeat) public {
          blacklists[_type]=BlacklistDetails(_start,_end,_repeat);
          inverstorToBlacklist[_investor]=_type;
          blacklistToAddress[_type].push(_investor);
      }
     /**
      * @dev to edit the existing blacklist entry
      */
      function editEntry(string _type,address _investor,uint256 _start,uint256 _end,uint256 _repeat) isPresent(_investor) public {
          blacklists[_type]=BlacklistDetails(_start,_end,_repeat);
          inverstorToBlacklist[_investor]=_type;
      }
     

}`

**Comment by kpulkit29** [comment]
This is a very bare bones logic I made.A lot of functionality needs to be added here.In the canBeTransferred function the startDate ,endDate is updated and also checked for blacklisting period

**Comment by kpulkit29** [comment]
Kindly tell if I am going in the right direction with this so that I can continue integrating it with other modules and writing truffle test for it

**Comment by satyamakgec** [comment]
@kpulkit29 Please first read the existing modules that are present in the [modules folder](https://github.com/PolymathNetwork/polymath-core/tree/development-1.5.0/contracts/modules) and understand the nomenclature that we are using. As mentioned in the above specs `verifyTransfer()` is the key function to check whether the transfer will be successfull or not and for transfer managers you should have to inherit the `ITransferManager.sol`. Please create the PR (WIP) so we can easily review it because it is impossible to us to give you a green flag on your apporach until unless you are following the same standards that we have in our ecosystem. 
If you need to understand i.e `how you can create polymath compatible modules ?` then please have a look into our [developer portal](https://developers.polymath.network/docs/doc4.html)

**Comment by kpulkit29** [comment]
Okay I will look into it

**Comment by pabloruiz55** [comment]
Hi @kpulkit29, I see a few problems specifically with the canBeTransferred function.
The logic seems off. If someone were to call it after a long time (let's say it repeats every 30 days and I only call this function after one year), the logic breaks as it will only try to sum the repeat once.

IMO, there's a way to do this that doesn't require saving state where you use division to know how many "iterations" have passed and add that difference to the start end date. 

Let's say my data is: 
```
start = 500
end = 1000
repeats = 2000 
now = 9700
```
The above means that my transfers would be blocked on:
* 500 to 1000
* 3000 to 3500
* 5500 to 6000
* 8000 to 8500
* 10500 to 11000
* and so on

Some pseudo-code:
```
int repetitions = floor(now / (end - start + repeats))  (equals 3 in my example above)
int newStart = start + ((end - start + repeats) * repetitions)
int newEnd = .end + ((end - start + repeats) * repetitions)

if (now > newStart && now < newEnd) -> rejectTransfer
```

**Comment by kpulkit29** [comment]
Thanks for the help.I would integrate with other modules and make a PR for
you guys to review

**Comment by kpulkit29** [comment]
Have made new PR https://github.com/PolymathNetwork/polymath-core/pull/276

**Comment by kpulkit29** [comment]
New blacklisting contract added Please explain what _perm means here in GeneralPermissionsManager.sol
function checkPermission(address _delegate, address _module, bytes32 _perm) external view returns(bool) {
if (delegateDetails[_delegate] != bytes32(0)) {
return perms[_module][_delegate][_perm];
}else
return false;
}
Also module address here would be address of blacklisting contract right??
Please review the contract.

**Comment by satyamakgec** [comment]
@kpulkit29 left the bunch of comments on the PR.

**Comment by kpulkit29** [comment]
@satyamakgec  I am making the changes you mentioned.So a new AutomatedBlacklistingFactory.sol also is required for deploying it right as different STO owners would like to make use of different modules

**Comment by CPSTL** [comment]
@kpulkit29 Hello! After discussing internally, we are unfortunately going to have to remove you from this issue/task. This task requires a more advanced skillset and we have a strict timeline to get it completed by. Please standby for other development tasks that you may be interested in. Thanks for your understanding :)

**Comment by kpulkit29** [comment]
I respect your decision.But I request you to give it back to me as I have been reading your contracts and I was close to get the contract up and running.I had invested a lot of time understanding the code.Also getting removed from a bounty has a really negative effect on my profile.

**Comment by kpulkit29** [comment]
I promise to deliver it 1-2 days

**Comment by kpulkit29** [comment]
@CPSTL Sir I think I deserved to be warned about getting removed rather than straightaway being removed

**Comment by kpulkit29** [comment]
https://github.com/PolymathNetwork/polymath-core/pull/279

**Comment by satyamakgec** [comment]
@kpulkit29 I appreciate your continuous efforts to this bounty, and I really like that you are a hardworking and curious guy. 
Anyways this bounty is not just for the coding Pulkit through this bounty we are trying to find the right candidate who will work with us in near future. So not only we are judging the coding skills but the strategy to solve a problem is also important. 
Don't be upset we are going to launch many more bounties (focused on coding) where you can try your best and collect the bounty. 
Cheers! :)
Satyam

**Comment by vpanwar821** [comment]
Hi,

I have made the PR [#289](https://github.com/PolymathNetwork/polymath-core/pull/289). Please review that and provide me the feedback. <br/>
Also,I need some more information about the delete blacklist type operation. As,mentioned in the requirements that approved staff can perform the delete operation so how this will work whether <br/>
```
it will delete the corresponding addresses with the blacklist type or
we need to restrict to delete when blacklist type have some addresses associated.
```
Thanks

**Comment by satyamakgec** [comment]
@vpanwar821 I don't think so we have to delete the whole blacklist type when addresses are associated with it. An issuer has the option to delete the blacklist for a particular `investor/affiliate` & An issuer can remove the blacklist type only when no address is associated with it.

**Comment by vpanwar821** [comment]
@satyamakgec Thanks for the information.

**Comment by gitcoinbot** [comment]
@vpanwar821 Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=1) | [3 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=3) | [5 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=5) | [10 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=10) | [100 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=100)

**Comment by vpanwar821** [comment]
I have submitted the PR [#289](https://github.com/PolymathNetwork/polymath-core/pull/289) and its review is in process.

**Comment by tanpv** [comment]
hi, just another algorithm to check now is inside lock time or not. I demo in js but will be very easy to transfer to sol.

const start = 1000
const end = 3000
const repeatPeriod = 4000
const now = 10000

const stepFromStartToNearestWithNow = Math.floor(((now-start) / repeatPeriod))
const startNearestwithNow = stepFromStartToNearestWithNow * repeatPeriod + start

const stepFromEndToNearestWithNow = Math.floor(((now-end) / repeatPeriod))
const endNearestwithNow = stepFromEndToNearestWithNow * repeatPeriod + end

if (startNearestwithNow > endNearestwithNow)
{
    console.log("now is inside lock")
} else {
    console.log("now is outside log")
}

**Comment by gitcoinbot** [comment]
@vpanwar821 Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=1) | [3 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=3) | [5 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=5) | [10 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=10) | [100 days](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246?snooze=100)

**Comment by vpanwar821** [comment]
Yes,I have submitted the PR #289 and its review is in process.

**Comment by CPSTL** [comment]
Thanks @vpanwar821 for your patience here. We are still reviewing.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 5000.0 POLY (870.0 USD @ $0.17/POLY) has been submitted by__:

 1. [@vpanwar821](https://gitcoin.co/profile/vpanwar821)



@CPSTL please take a look at the submitted work:
* [PR](https://github.com/PolymathNetwork/polymath-core/pull/289) by @vpanwar821

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $41,585.52 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 5000.0 POLY (870.0 USD @ $0.17/POLY)  attached to this issue has been approved & issued to @vpanwar821.__  

 * Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/PolymathNetwork/polymath-core/272/1246)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $49,667.37 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by satyamakgec** [comment]
Closing in the favor of #289
